lib/posix-process: Supplementary changes for multiprocess

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

⚠️ **Deprecation notice:**

* CONFIG_LIBPOSIX_PROCESS_PIDS
* CONFIG_LIBPOSIX_PROCESS_INIT_PIDS
* CONFIG_LIBPOSIX_PROCESS_CLONE
* CONFIG_LIBPOSIX_PROCESS_VFORK 

The following are consolidated into a single option, CONFIG_LIBPOSIX_PROCESS_MULTITHREADING:
* CONFIG_LIBPOSIX_PROCESS_PIDS
* CONFIG_LIBPOSIX_PROCESS_CLONE

The following is replaced by CONFIG_LIBPOSIX_PROCESS_MULTIPROCESS:
* CONFIG_LIBPOSIX_PROCESS_VFORK 

CONFIG_LIBPOSIX_PROCESS_INIT_PIDS is deprecated, and its functionality is enabled by default.

### Description of changes

This series consists of a set of supplementary changes emerging from the work on multiprocess. These consist of fixes as well as major refactoring of libposix-process. In brief:
* Refactor Kconfig options of libposix-process for multithreading / multiprocess: This consolidates exiting options into more meaningful ones and removes redundancy. Affects both internal and external libraries.
* Make process management functions internal: Removes redundant functions from the public API of libposix-process.
* Assign a fixed pid to init: Deprecates CONFIG_LIBPOSIX_PROCESS_INIT_PIDS and makes its functionality enabled by default.
* Restrict pthread creation to clone(): This fixes a long standing issue where new kernel threads will always be associated to new pthreads.
* Various fixes

For details see the commit messages of individual PRs.

Depends-on: https://github.com/unikraft/app-elfloader/pull/90
Depends-on: https://github.com/unikraft/lib-musl/pull/87

michpappas committed:
lib/posix-process: Deprecate POSIX_PROCESS_INIT_PIDS

Retire POSIX_PROCESS_INIT_PIDS and assign PID 1 to the init process by
default. This improves consistency and reduces configuration complexity.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Refactor Kconfig options

Replace CONFIG_LIBPOSIX_PROCESS_PIDS and CONFIG_LIBPOSIX_PROCESS_CLONE
with a single option, CONFIG_LIBPOSIX_PROCESS_MULTITHREADING. This
option enables clone(), _exit(), and exit_group() syscalls as well as
the core logic of libposix-process. Make options previously relevant
to CONFIG_LIBPOSIX_PROCESS_CLONE move to subselection of multithreading.

Without CONFIG_LIBPOSIX_PROCESS_MULTITHREADING the implementation falls
back into stubs. This provides an execution environment for simple
libc-based applications that don't require multithreading, but still
need to execute under a process-like environment. Under this state,
the syscalls involved in multithreading (or multiprocess) return an
error, while others are emulated. Notice that if Unikraft is not configured
to execute main() in a separate uk_thread (i.e. LIBUKBOOT_MAINTHREAD),
_exit() and exit_group() are not available.

Introduce a new config option to enable multiprocess support, namely
CONFIG_LIBPOSIX_PROCESS_MULTIPROCESS. This option replaces the config
option of vfork(), and enforces the selection of multithreading and
execve().

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Add declaration of vfork()

Update unistd.h with declaration of vfork(). This is required when
writing applications with nolibc.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Fix assertion in execve()

Update assertion to check against UK_PID_INIT instead of zero.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Update semantics of vfork()'s return type

Although with the current implementation the returned value is identical,
vfork() returns the child's pid rather than tid. Updadte the naming of
identifiers to reflect that correctly.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/vfscore: Adapt to updated libposix-process config options

Following the consolidation of CONFIG_LIBPOSIX_PROCESS_PIDS and
CONFIG_LIBPOSIX_PROCESS_CLONE into LIBPOSIX_PROCESS_MULTITHREADING,
update to the new config.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-futex: Adapt to updated libposix-process config options

Following the consolidation of CONFIG_LIBPOSIX_PROCESS_PIDS and
CONFIG_LIBPOSIX_PROCESS_CLONE into LIBPOSIX_PROCESS_MULTITHREADING,
update to the new config.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Adapt to updated libposix-process config options

Following the consolidation of CONFIG_LIBPOSIX_PROCESS_PIDS and
CONFIG_LIBPOSIX_PROCESS_CLONE into LIBPOSIX_PROCESS_MULTITHREADING,
update to the new config.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/uklock: Adapt to updated libposix-process config options

Following the consolidation of CONFIG_LIBPOSIX_PROCESS_PIDS and
CONFIG_LIBPOSIX_PROCESS_CLONE into LIBPOSIX_PROCESS_MULTITHREADING,
update to the new config.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-fdtab: Adapt to updated libposix-process config options

Following the consolidation of CONFIG_LIBPOSIX_PROCESS_PIDS and
CONFIG_LIBPOSIX_PROCESS_CLONE into LIBPOSIX_PROCESS_MULTITHREADING,
update to the new config.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/syscall-shim: Adapt to updated libposix-process config options

Following the consolidation of CONFIG_LIBPOSIX_PROCESS_PIDS and
CONFIG_LIBPOSIX_PROCESS_CLONE into LIBPOSIX_PROCESS_MULTITHREADING,
update to the new config.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
Regarding the failure of the libc-test: Should be related to the outdated options in libmusl.

andreittr commented on lib/posix-process/Config.uk:
```diff
@@ -4,50 +4,44 @@ menuconfig LIBPOSIX_PROCESS
 config LIBPOSIX_PROCESS_CLONE_PREFER_CHILD
 	bool "Prefer scheduling of child"
-	depends on LIBPOSIX_PROCESS_CLONE
+	depends on LIBPOSIX_PROCESS_MULTITHREADING
```
This `depends on` line might be redundant, as the config option is already in an `if LIBPOSIX_PROCESS_MULTITHREADING` block.

andreittr commented on lib/posix-process/Config.uk:
```diff
@@ -4,50 +4,44 @@ menuconfig LIBPOSIX_PROCESS
 
 config LIBPOSIX_PROCESS_MAX_PID
-	int "Largest PID"
+	int "Largest TID"
```
I understand why the label change to "TID" (and why keep the config var the same); would it perhaps be more descriptive to users to say "Largest TID/PID" instead? Or do you think there would be more confusion if we mention PIDs for an option that pops up without multiprocessing enabled?

michpappas replied:
> Or do you think there would be more confusion if we mention PIDs for an option that pops up without multiprocessing enabled?

Exactly. And in addition, with multiprocess enabled one would expect to find it under that menu instead, so I opted to move it under multithreading and rename it to something sensible, and hope it would be intuitive to the user that when multiprocess is enabled it also implies max pid. Saying that, I don't like it very much either, so I'm open to suggestions.

andreittr replied:
> I'm open to suggestions.

Only thing I can think of are 2 separate Kconfig options: "Largest TID" and "Largest TID/PID", first selected if `multithread && !multiproc`, 2nd if `multiproc`; then some Kconfig magic to turn these 2 vars into a single var that code uses.

Or, y'know, just keep TID, user is configuring a kernel, they should know what's up :smile:

andreittr commented on lib/posix-process/include/uk/process.h:
```diff
@@ -43,7 +43,7 @@
 #include <uk/thread.h>
 #endif
```
AFAICT (pls correct me if I'm wrong), `<uk/thread.h>` is used by multithreading code only (and the multithread Kconfig option ensures uksched is selected).
I believe we could move this `#include` into the `#if CONFIG_LIBPOSIX_PROCESS_MULTITHREADING` block lower down.

michpappas replied:
That's right, the only thing is that without it we get a compile error because of `NULL` not being defined. I think `thread.h` resolves that by some chain include that uses `stddef.h`. Anyway, I'll move `thread.h` inside the `CONFIG_LIBPOSIX_PROCESS_MULTITHREADING` block, and include `stddef.h` for now, and at some point push another commit to replace boolean types and NULL with the equivalent kernel types.

andreittr commented on lib/posix-process/process.c:
```diff
@@ -749,7 +752,25 @@ static int pprocess_clone_thread(const struct clone_args *cl_args __unused,
 				 struct uk_thread *child __unused,
 				 struct uk_thread *parent __unused)
```
`child`, `parent`, and `cl_args` are now used, you can remove the `__unused` attr.

andreittr commented on lib/posix-fdtab/Config.uk:
```diff
@@ -13,7 +13,7 @@ if LIBPOSIX_FDTAB
 	config LIBPOSIX_FDTAB_MULTITAB
 	bool
-	default y if LIBPOSIX_PROCESS_CLONE
+	default y if LIBPOSIX_PROCESS_MULTITHREADING
```
The multitab logic could (in theory) now be constrained to only `if LIBPOSIX_PROCESS_MULTIPROCESS`, and the compat stop-gaps removed, but that's a change in behavior outside the scope of this work.
Nevertheless, something to consider soon.

andreittr reviewed (commented):
Change looks nice, I like the featureset approach to `posix-process` config.
I've just some minor remarks.

mogasergiu commented on lib/posix-process/Config.uk:
```diff
@@ -4,50 +4,44 @@ menuconfig LIBPOSIX_PROCESS
 
 if LIBPOSIX_PROCESS
 
-menuconfig LIBPOSIX_PROCESS_PIDS
```
Commit meassage headline for 
 lib/posix-process: Refactor Kconfig options for multithreading/multip…

…rocess

is too long and spills into the next line

I know you added a Checkpatch-Ignore, but I'd rather not ahve it like this in our upstream history.

mogasergiu commented on lib/posix-process/clone.c:
```diff
@@ -515,15 +515,6 @@ int uk_clone(struct clone_args *cl_args, size_t cl_args_len,
 		    (child->tlsp != child->uktlsp) ? "custom" : "Unikraft");
 
-	if (!(cl_args->flags & CLONE_THREAD)) {
-		uk_pr_debug("Creating new process for child\n");
```
In commit message of ` lib/posix-process: Restrict pthread creation to clone() `: `s/rename then/rename them/`

michpappas replied:
This commit is now gone.

mogasergiu commented on lib/posix-process/include/uk/process.h:
```diff
@@ -218,6 +211,15 @@ struct uk_posix_clonetab_entry {
+ * support paths and therefore we cannot execve(). For more details
+ * see the implementation of app-elfloader.
+ */
+int uk_posix_process_create_pthread(struct uk_thread *thread);
```
Why don't we move it to elf_binfmt in elfloader then?

mogasergiu commented on lib/posix-process/vfork.c:
```diff
@@ -18,21 +18,17 @@
-	struct posix_process *child_proc;
 	struct clone_args cl_args = {0};
-	pid_t child_tid;
+	pid_t pid; /* child */
```
Commit subject is a bit misleading, maybe rewrite as to imply it refers to fixing nomenclature rather than C-type

michpappas replied:
Good point, will update 👍🏼

mogasergiu commented on lib/posix-process/Config.uk:
```diff
@@ -4,50 +4,44 @@ menuconfig LIBPOSIX_PROCESS
+menuconfig LIBPOSIX_PROCESS_MULTITHREADING
+	bool "Multithreading support"
 	select LIBUKSCHED
+	select LIBSYSCALL_SHIM
```
Why do we need syscall shim as mandatory now?

michpappas replied:
Okay so these are macros that are always compiled in right. Will update, thanks 👍🏼

michpappas committed:
lib/posix-process: Ignore CLONE_VM when assigning tls parameter

The interpretation of the tls parameter of clone() is conditional to
passing CLONE_SETTLS in flags. Fix a bug where CLONE_VM would result
into assigning the tls parameter to child. Assigning the parent's tls
when CLONE_VFORK is set, is performed in clone_setup_child_ctx(),
conditionally that the child's tls is not already set, presumably due
to CLONE_SETTLS.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Consolidate assignment of pthread parent and self

uk_posix_process_create() is missing the assignment of pthread_self.
Instead of having wrappers perform these assignments of parent and
self, move these assignments to the core functions that create and
release threads.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
Rebased to the latest state of #1618

michpappas commented:
Addressed review comments, added #1626 into the dependencies to avoid a conflict, pushed an additional patch to fix a bug in `clone()`.

michpappas commented:
@andreittr, @mogasergiu I pushed a bunch updates. The TLDR is:
* Rebased to the latest state of #1618
* What https://github.com/unikraft/unikraft/pull/1627#issuecomment-2783362514 says
* The cummulative diff of chagnes excl. the rebase: https://github.com/unikraft/unikraft/compare/88b066ae8cf42f9ce18197925770a1898191baaf..1ab140af44a880ed24c4d204e99ec5aced775d16
* libc test fails because the relevant change in musl pending merge

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

michpappas committed:
lib: Add Kconfig option to signify support of VFS

Add HAVE_VFS option in Kconfig for features that depend on VFS support,
and remove redundant defaults. Update libvfscore to select HAVE_VFS.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Add uk_posix_process_create_pthread()

Provide a function to attach a thread to the current process. This is
intended to be exclusively used by app-elfloader when configured with
initrd, in order to assign the container thread it creates to the init
process.

Move uk_process_kill() to the internal API, and rename to avoid using
the naming convention of public functions, and deprecate the unused
uk_posix_process_create().

Checkpatch-Ignore: REPEATED_WORD
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Restrict pthread creation to clone()

The current implementation incorrectly creates a pthread in the current
process, for every new uk_thread created. This behavior is problematic
because during a syscall, libraries may create uk_threads that are
intended to act as kernel threads. Decouple pthread initializaiton from
uk_thread and limit it to clone().

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

andreittr commented on lib/Config.uk:
```diff
@@ -7,31 +7,27 @@
-       default n
+
+config HAVE_VFS
+	bool
```
Thanks for this, it'll make the transition to `posix-vfs` smoother :pray: .

andreittr commented on lib/posix-process/Config.uk:
```diff
@@ -4,35 +4,17 @@ menuconfig LIBPOSIX_PROCESS
-	select LIBUKSCHED
-
-if LIBPOSIX_PROCESS_PIDS
+menuconfig LIBPOSIX_PROCESS_MULTITHREADING
```
Typo in the "Refactor Kconfig options" commit message, last paragraph: multirpocess -> multiprocess

andreittr commented on lib/posix-process/Makefile.uk:
```diff
@@ -38,24 +37,22 @@ LIBPOSIX_PROCESS_SRCS-y += $(LIBPOSIX_PROCESS_BASE)/signal/kill.c
-ifeq ($(CONFIG_LIBPOSIX_PROCESS_CLONE), y)
+ifeq ($(CONFIG_LIBPOSIX_PROCESS_MULTITHREADING), y)
 LIBPOSIX_PROCESS_SRCS-y += $(LIBPOSIX_PROCESS_BASE)/signal/clone.c|signal
 endif
```
Is there a reason I'm not seeing for keeping this an `ifeq` block rather than `LIBPOSIX_PROCESS_SRCS-$(CONFIG_LIBPOSIX_PROCESS_MULTITHREADING)`?

michpappas replied:
I think it's how it evolved once making it unconditional due to the stubs. Will push an update in #1628 and update here accordingly, thanks.

andreittr commented on lib/syscall_shim/Makefile.uk:
```diff
@@ -100,7 +100,7 @@ LIBSYSCALL_SHIM_LIBC_STUBS_FLAGS-$(call have_gcc) += -Wno-builtin-declaration-mi
 
 LIBSYSCALL_SHIM_SRCS-y += $(LIBSYSCALL_SHIM_BASE)/uk_prsyscall.c
-LIBSYSCALL_SHIM_SRCS-$(CONFIG_LIBPOSIX_PROCESS_EXECVE) += $(LIBSYSCALL_SHIM_BASE)/execve.c
+LIBSYSCALL_SHIM_SRCS-$(CONFIG_LIBPOSIX_PROCESS_MULTIPROCESS) += $(LIBSYSCALL_SHIM_BASE)/execve.c
```
Since `LIBPOSIX_PROCESS_EXECVE` remains a separate option, I think this change should be dropped here, as we want the `execve` syscall even if `!multiprocess`.

michpappas replied:
Ooops, I totally forgot to update these, thanks for the catch. Will also update the commit message to remove the paragraph about execve.

andreittr commented on lib/posix-fdtab/fdtab.c:
```diff
@@ -28,14 +28,14 @@
 #include <uk/event.h>
 #include <uk/prio.h>
-#endif /* CONFIG_LIBPOSIX_PROCESS_EXECVE */
+#endif /* CONFIG_LIBPOSIX_PROCESS_MULTIPROCESS */
```
fdtab in particular should have its cloexec handler run on any exec, regardless of multiproc or not; I think this should remain `CONFIG_LIBPOSIX_PROCESS_EXECVE` (and similarly around the handler).

michpappas replied:
Yup, same as above, forgot to update.

andreittr reviewed (commented):
Looks very good; did a full pass and have only one issue with execve vs. multiproc `ifdefs` in libs.

andreittr reviewed (approved):
Looks great!

Reviewed-by: Andrei Tatar <andrei@unikraft.io>

michpappas commented:
Rebased from `staging`

andreittr commented:
Approved-by: Andrei Tatar <andrei@unikraft.io>
